Formal2024/Workshops/workshop4.lean

```diff
@@ -0,0 +1,251 @@
+-- elements satisfying `P`
+
+#check subset_def
+#check mem_setOf
+#check inter_def
+#check union_def
+#check mem_diff
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  simp only [subset_def, inter_def, inter_def, mem_setOf]
+  rw [subset_def] at h
+  -- We can use `simp only` or `rw` to unfold the definitions
+  -- `simp only` is a more powerful `rw`, that also works under quantifiers
+  -- Now we can proceed as usual
+  rintro x ⟨xs, xu⟩
+  exact ⟨h x xs, xu⟩
+  tada
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  -- This is not needed though, as the anonymus constructor forces Lean to unfold the definitions
+  rintro x ⟨xs, xu⟩
+  exact ⟨h xs, xu⟩ -- There may be some slight differences (here we do not need to specify x)
+  tada
+
+
+example : s ∩ (t ∪ u) ⊆ s ∩ t ∪ s ∩ u := by
+  intro x hx                      -- Subset works like an implication
+  have xs : x ∈ s := hx.1         -- Intersection works like an `∧`
+  have xtu : x ∈ t ∪ u := hx.2    -- Union works like an `∨`
+  rcases xtu with xt | xu
+  · left
+    show x ∈ s ∩ t
+    exact ⟨xs, xt⟩
+  . right
+    show x ∈ s ∩ u
+    exact ⟨xs, xu⟩
+  tada
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  intro x xstu
+  have xs : x ∈ s := xstu.1.1   -- `x ∈ s \ t` means `x ∈ s ∧ x ∉ t`
+  have xnt : x ∉ t := xstu.1.2
+  have xnu : x ∉ u := xstu.2
+  constructor
+  · exact xs
+  intro xtu
+  -- x ∈ t ∨ x ∈ u
+  rcases xtu with xt | xu
+  · show False; exact xnt xt
+  . show False; exact xnu xu
+  tada
+
+-- This is quite lengthy. We can use `rintro` to shorten the proof
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  rintro x ⟨⟨xs, xnt⟩, xnu⟩
+  use xs
+  rintro (xt | xu) <;> contradiction -- We often need brackets when using `rintro` on a conjunction
+  tada
+
+-- Two sets are equal when they have the same elements. This is called "extensionality"
+-- You can use it by typing `ext _`
+
+example : s ∩ t = t ∩ s := by
+  ext x -- You can give the variable a name by typing it instead of `_`
+  constructor
+  · rintro ⟨xs, xt⟩; exact ⟨xt, xs⟩
+  . rintro ⟨xt, xs⟩; exact ⟨xs, xt⟩
+  tada
+
+-- A similar statement is the theorem `Subset.antisymm`
+
+#check Subset.antisymm
+
+-- Membership to intersection and union can be unfolded using the following two theorems
+-- (although Lean usually does not need it)
+
+#check mem_union
+#check mem_inter_iff
+
+-- Exercises (MIL 4.1)
+
+-- 1
+example : s ∩ (s ∪ t) = s := by
+  sorry
+
+-- 2
+example : s ∪ s ∩ t = s := by
+  sorry
+
+-- 3
+example : s \ t ∪ t = s ∪ t := by
+  sorry
+
+-- 4
+example : s \ t ∪ t \ s = (s ∪ t) \ (s ∩ t) := by
+  sorry
+
+-- 5
+example : s ∩ t ∪ s ∩ u ⊆ s ∩ (t ∪ u) := by
+  sorry
+
+end
+
+/-
+The main difference between sets and types is the predicate `∈`. In fact, in Lean a set coincides
+with its `∈` predicate:
+
+`Set α` is defined as `α → Prop`. If `s : Set α` and `a : α`, then `a ∈ s` is defined as `s a`.
+The set `{ y | P y }` is defined as the function `fun y ↦ P y` and `x ∈ { y | P y }` means `P x`.
+-/
```

Comment:
This is true, but sometimes Lean gets a little confused (I don't know why) and forgets about this. If `h : x ∈ { y | P y }` and Lean refuses to recognise `h : P x`, then `h.out` definitely does the job.